pkgs/by-name/bi/bitwarden-directory-connector/package.nix

```diff
@@ -0,0 +1,68 @@
+{
+  lib,
+  buildNpmPackage,
+  fetchFromGitHub,
+  pkgs,
+  jq,
+  python3,
+  pkg-config,
+  libsecret,
+  nodejs_18,
+}: let
+  buildNpmPackage' = buildNpmPackage.override {nodejs = nodejs_18;};
+in
+  buildNpmPackage' rec {
+    pname = "bitwarden-directory-connector";
+    version = "2023.10.0";
+
+    src = fetchFromGitHub {
+      owner = "bitwarden";
+      repo = "directory-connector";
+      rev = "v${version}";
+      hash = "sha256-PlOtTh+rpTxAv8ajHBDHZuL7yeeLVpbAfKEDPQlejIg=";
+    };
+
+    postPatch = ''
+      jq 'del(.scripts.preinstall)' package.json > package.json.tmp
```

Comment:
I'm okay with allowing `nativeBuildInputs` in the fetcher, though it might be better if they were after `prefetch-npm-deps`. We could also just copy the structure from how the rust fetcher does it as well for something like this maybe?

```diff
diff --git a/pkgs/build-support/node/fetch-npm-deps/default.nix b/pkgs/build-support/node/fetch-npm-deps/default.nix
index 725f9ba3bb01..0e60bdd2d346 100644
--- a/pkgs/build-support/node/fetch-npm-deps/default.nix
+++ b/pkgs/build-support/node/fetch-npm-deps/default.nix
@@ -151,6 +151,7 @@
     , hash ? ""
     , forceGitDeps ? false
     , forceEmptyCache ? false
+    , nativeBuildInputs ? []
     , ...
     } @ args:
     let
@@ -168,7 +169,7 @@
     stdenvNoCC.mkDerivation (args // {
       inherit name;
 
-      nativeBuildInputs = [ prefetch-npm-deps ];
+      nativeBuildInputs = [ prefetch-npm-deps ] ++ nativeBuildInputs;
 
       buildPhase = ''
         runHook preBuild
```

I'm a bit hesitant to forward `nativeBuildInputs` into the fetcher from `buildNpmPackage` by default though because setup hooks might do weird things, so I might be more comfortable copying what `buildRustPackage` does with that as well and add something like their `depsExtraArgs` to the builder function

```diff
diff --git a/pkgs/build-support/node/build-npm-package/default.nix b/pkgs/build-support/node/build-npm-package/default.nix
index 42c6a9c065b2..0002ae9b8987 100644
--- a/pkgs/build-support/node/build-npm-package/default.nix
+++ b/pkgs/build-support/node/build-npm-package/default.nix
@@ -44,11 +44,12 @@
   # Value for npm `--workspace` flag and directory in which the files to be installed are found.
 , npmWorkspace ? null
 , nodejs ? topLevelArgs.nodejs
-, npmDeps ?  fetchNpmDeps {
+, depsExtraArgs ? {}
+, npmDeps ?  fetchNpmDeps ({
   inherit forceGitDeps forceEmptyCache src srcs sourceRoot prePatch patches postPatch;
   name = "${name}-npm-deps";
   hash = npmDepsHash;
-}
+} // depsExtraArgs)
 , ...
 } @ args:
 
```

(Unrelated but ugh I wish the `fetchNpmDeps` call wasn't confusingly moved into the argument list. I'm tempted to make the PR to allow `nativeBuildInputs` myself so that I can move that back somewhere more reasonable that doesn't give my brain trouble parsing it)

Reply:
An example of how to use depsExtraArgs can be found here https://github.com/NixOS/nixpkgs/blob/86b7775ff38a6f12529691d8bc39ffb7d219f5a2/pkgs/development/tools/misc/samply/default.nix#L24-L28

Reply:
For this case it would also be fine to use `preBuild` instead, I think, since that isn't propagated to the fetcher, and the fetcher doesn't need the change anyway.

Reply:
Could this be why it recently failed to build on master for me? https://github.com/NixOS/nixpkgs/pull/265783#issuecomment-1864724367

(also didnt expect any activity today of all days)

Reply:
Yes, because you have the FOD already cached locally but it can now no longer be reproduced when stdenv changes.

Reply:
Might it be better to not pass in jq at all, instead to call it like ``${lib.getExe jq}`` isntead of ``jq``?  
Since its technically not needed to build the application iteself but ratehr a tool used to make the install flow easier.

Reply:
Why would it have to move?

What I mean is to have it as:
```
    postPatch = ''
      ${lib.getExe jq} 'del(.scripts.preinstall)' package.json > package.json.tmp
      mv -f package.json{.tmp,}
    '';
```

Just changing ``postPatch`` to ``preBuild`` yields this:

```
[ERROR] Failed to build kitt - Last 20 lines of logs:
[ERROR]   stderr) node_modules/rimraf/bin.js: interpreter directive changed from "#!/usr/bin/env node" to "/nix/store/if3abix5p1mnlqabshwyj4hk3fyf7cwq-nodejs-18.19.0/bin/node"
[ERROR]   stderr) node_modules/esbuild/bin/esbuild: interpreter directive changed from "#!/usr/bin/env node" to "/nix/store/if3abix5p1mnlqabshwyj4hk3fyf7cwq-nodejs-18.19.0/bin/node"
[ERROR]   stderr) node_modules/eslint-config-prettier/bin/cli.js: interpreter directive changed from "#!/usr/bin/env node" to "/nix/store/if3abix5p1mnlqabshwyj4hk3fyf7cwq-nodejs-18.19.0/bin/node"
[ERROR]   stderr) node_modules/js-yaml/bin/js-yaml.js: interpreter directive changed from "#!/usr/bin/env node" to "/nix/store/if3abix5p1mnlqabshwyj4hk3fyf7cwq-nodejs-18.19.0/bin/node"
[ERROR]   stderr) npm ERR! code 127
[ERROR]   stderr) npm ERR! path /build/source
[ERROR]   stderr) npm ERR! command failed
[ERROR]   stderr) npm ERR! command sh -c npm run sub:init
[ERROR]   stderr) npm ERR! > @bitwarden/directory-connector@0.0.0 sub:init
[ERROR]   stderr) npm ERR! > git submodule update --init --recursive
[ERROR]   stderr) npm ERR! sh: line 1: git: command not found
[ERROR]   stderr)
[ERROR]   stderr) npm ERR! A complete log of this run can be found in: /build/cache/_logs/2023-12-28T01_19_25_763Z-debug-0.log
[ERROR]   stderr)
[ERROR]   stderr) error: builder for '/nix/store/8zkp8kiyc0bnid2x1ij06fmz4acbdvsd-bitwarden-directory-connector-2023.10.0.drv' failed with exit code 127
[ERROR]   stderr) error: 1 dependencies of derivation '/nix/store/6bdjylakh6ri9vgaq37mbgwz9jqf1ixp-unit-bitwarden-directory-connector.service.drv' failed to build
[ERROR]   stderr) error: 1 dependencies of derivation '/nix/store/5wdms3wzqbisx3x4fwgnhabxzak6imas-system-units.drv' failed to build
[ERROR]   stderr) error: 1 dependencies of derivation '/nix/store/xil8g7ji4gs0wm7vdvay988lj0b9mjnr-etc.drv' failed to build
[ERROR]   stderr) error: 1 dependencies of derivation '/nix/store/1lzv8h4h1bdxiqfh2hkbjp9sll62a8g1-nixos-system-unnamed-24.05pre-git.drv' failed to build
[ERROR]  failure) Child process exited with error code: 100
[ERROR] -----
[ERROR] Operation failed with error: Child process exited with error code: 100
```

Reply:
> What I mean is to have it as:

That would also work if we use buildPackages.jq instead. Good idea!

Reply:
Will I go for that then?

Reply:
Sure, go ahead. I think we reviewed and tested so much in this PR already, that anything else can then be fixed in a new one, to get this finally merged.